File: RadixCyclicNN/go/radixnet/mt19937.go
package radixnet

// A Mersenne Twister with the exact semantics of CPython's random.Random:
// the same seeding (init_by_array over the 32-bit chunks of |seed|), the same
// 53-bit doubles and the same state layout, so an rng_state written by either
// implementation continues the other's sequence.

import "fmt"

const (
	mtN         = 624
	mtM         = 397
	mtMatrixA   = 0x9908b0df
	mtUpperMask = 0x80000000
	mtLowerMask = 0x7fffffff
)

// MT19937 is a Python-compatible pseudo random generator.
type MT19937 struct {
	mt    [mtN]uint32
	index int
}

// NewMT19937 seeds a generator like random.Random(seed).
func NewMT19937(seed int64) *MT19937 {
	r := &MT19937{}
	r.Seed(seed)
	return r
}

// Seed reseeds like random.seed(seed) for an integer seed: the key is |seed|
// split into 32-bit words, least significant first (a zero seed gives [0]).
func (r *MT19937) Seed(seed int64) {
	n := seed
	if n < 0 {
		n = -n
	}
	key := []uint32{}
	u := uint64(n)
	for {
		key = append(key, uint32(u&0xffffffff))
		u >>= 32
		if u == 0 {
			break
		}
	}
	r.initByArray(key)
}

func (r *MT19937) initGenrand(s uint32) {
	r.mt[0] = s
	for i := 1; i < mtN; i++ {
		r.mt[i] = 1812433253*(r.mt[i-1]^(r.mt[i-1]>>30)) + uint32(i)
	}
	r.index = mtN
}

func (r *MT19937) initByArray(key []uint32) {
	r.initGenrand(19650218)
	i, j := 1, 0
	k := mtN
	if len(key) > k {
		k = len(key)
	}
	for ; k > 0; k-- {
		r.mt[i] = (r.mt[i] ^ ((r.mt[i-1] ^ (r.mt[i-1] >> 30)) * 1664525)) + key[j] + uint32(j)
		i++
		j++
		if i >= mtN {
			r.mt[0] = r.mt[mtN-1]
			i = 1
		}
		if j >= len(key) {
			j = 0
		}
	}
	for k = mtN - 1; k > 0; k-- {
		r.mt[i] = (r.mt[i] ^ ((r.mt[i-1] ^ (r.mt[i-1] >> 30)) * 1566083941)) - uint32(i)
		i++
		if i >= mtN {
			r.mt[0] = r.mt[mtN-1]
			i = 1
		}
	}
	r.mt[0] = 0x80000000
	r.index = mtN
}

// Uint32 returns the next 32-bit output (genrand_uint32).
func (r *MT19937) Uint32() uint32 {
	if r.index >= mtN {
		var kk int
		for kk = 0; kk < mtN-mtM; kk++ {
			y := (r.mt[kk] & mtUpperMask) | (r.mt[kk+1] & mtLowerMask)
			r.mt[kk] = r.mt[kk+mtM] ^ (y >> 1) ^ (mtMatrixA * (y & 1))
		}
		for ; kk < mtN-1; kk++ {
			y := (r.mt[kk] & mtUpperMask) | (r.mt[kk+1] & mtLowerMask)
			r.mt[kk] = r.mt[kk+mtM-mtN] ^ (y >> 1) ^ (mtMatrixA * (y & 1))
		}
		y := (r.mt[mtN-1] & mtUpperMask) | (r.mt[0] & mtLowerMask)
		r.mt[mtN-1] = r.mt[mtM-1] ^ (y >> 1) ^ (mtMatrixA * (y & 1))
		r.index = 0
	}
	y := r.mt[r.index]
	r.index++
	y ^= y >> 11
	y ^= (y << 7) & 0x9d2c5680
	y ^= (y << 15) & 0xefc60000
	y ^= y >> 18
	return y
}

// Float64 returns the next double in [0, 1) exactly like random.random().
func (r *MT19937) Float64() float64 {
	a := r.Uint32() >> 5
	b := r.Uint32() >> 6
	return (float64(a)*67108864.0 + float64(b)) * (1.0 / 9007199254740992.0)
}

// Uniform returns a + (b - a) * random(), like random.uniform.
func (r *MT19937) Uniform(a, b float64) float64 { return a + (b-a)*r.Float64() }

// State returns the generator state in Python's getstate() layout:
// [3, [624 words..., index], null].
func (r *MT19937) State() []any {
	words := make([]any, mtN+1)
	for i, w := range r.mt {
		words[i] = uint64(w)
	}
	words[mtN] = r.index
	return []any{3, words, nil}
}

// SetState restores a state produced by State() or by Python's getstate().
func (r *MT19937) SetState(state []any) error {
	if len(state) < 2 {
		return fmt.Errorf("rng_state must have at least 2 entries")
	}
	words, ok := state[1].([]any)
	if !ok || len(words) != mtN+1 {
		return fmt.Errorf("rng_state must hold %d words plus the index", mtN)
	}
	var mt [mtN]uint32
	for i := 0; i < mtN; i++ {
		v, err := anyToUint64(words[i])
		if err != nil || v > 0xffffffff {
			return fmt.Errorf("rng_state word %d is not a 32-bit integer", i)
		}
		mt[i] = uint32(v)
	}
	idx, err := anyToUint64(words[mtN])
	if err != nil || idx > mtN {
		return fmt.Errorf("rng_state index out of range")
	}
	r.mt = mt
	r.index = int(idx)
	return nil
}

func anyToUint64(v any) (uint64, error) {
	switch x := v.(type) {
	case float64:
		if x < 0 || x != float64(uint64(x)) {
			return 0, fmt.Errorf("not an unsigned integer: %v", x)
		}
		return uint64(x), nil
	case uint64:
		return x, nil
	case int:
		if x < 0 {
			return 0, fmt.Errorf("negative")
		}
		return uint64(x), nil
	case int64:
		if x < 0 {
			return 0, fmt.Errorf("negative")
		}
		return uint64(x), nil
	case uint32:
		return uint64(x), nil
	}
	return 0, fmt.Errorf("unexpected type %T", v)
}

File: RadixCyclicNN/rust/src/mt19937.rs
//! A Mersenne Twister with the exact semantics of CPython's `random.Random`:
//! the same seeding (`init_by_array` over the 32-bit chunks of `|seed|`), the
//! same 53-bit doubles and the same state layout, so a walk sampled here draws
//! the same numbers the Python and Go implementations draw.

const N: usize = 624;
const M: usize = 397;
const MATRIX_A: u32 = 0x9908_b0df;
const UPPER_MASK: u32 = 0x8000_0000;
const LOWER_MASK: u32 = 0x7fff_ffff;

/// A Python-compatible pseudo random generator.
#[derive(Clone)]
pub struct Mt19937 {
    mt: [u32; N],
    index: usize,
}

impl Mt19937 {
    /// Seeds a generator like `random.Random(seed)`.
    pub fn new(seed: i64) -> Mt19937 {
        let mut rng = Mt19937 { mt: [0; N], index: N };
        rng.seed(seed);
        rng
    }

    /// Reseeds like `random.seed(seed)` for an integer seed: the key is
    /// `|seed|` split into 32-bit words, least significant first.
    pub fn seed(&mut self, seed: i64) {
        let n = seed.unsigned_abs();
        let mut key = Vec::new();
        let mut u = n;
        loop {
            key.push((u & 0xffff_ffff) as u32);
            u >>= 32;
            if u == 0 {
                break;
            }
        }
        self.init_by_array(&key);
    }

    fn init_genrand(&mut self, s: u32) {
        self.mt[0] = s;
        for i in 1..N {
            let prev = self.mt[i - 1];
            self.mt[i] = 1812433253u32.wrapping_mul(prev ^ (prev >> 30)).wrapping_add(i as u32);
        }
        self.index = N;
    }

    fn init_by_array(&mut self, key: &[u32]) {
        self.init_genrand(19650218);
        let (mut i, mut j) = (1usize, 0usize);
        let mut k = N.max(key.len());
        while k > 0 {
            let prev = self.mt[i - 1];
            self.mt[i] = (self.mt[i] ^ (prev ^ (prev >> 30)).wrapping_mul(1664525))
                .wrapping_add(key[j])
                .wrapping_add(j as u32);
            i += 1;
            j += 1;
            if i >= N {
                self.mt[0] = self.mt[N - 1];
                i = 1;
            }
            if j >= key.len() {
                j = 0;
            }
            k -= 1;
        }
        k = N - 1;
        while k > 0 {
            let prev = self.mt[i - 1];
            self.mt[i] = (self.mt[i] ^ (prev ^ (prev >> 30)).wrapping_mul(1566083941)).wrapping_sub(i as u32);
            i += 1;
            if i >= N {
                self.mt[0] = self.mt[N - 1];
                i = 1;
            }
            k -= 1;
        }
        self.mt[0] = 0x8000_0000;
        self.index = N;
    }

    /// The next 32-bit output (`genrand_uint32`).
    pub fn next_u32(&mut self) -> u32 {
        if self.index >= N {
            let twist = |a: u32, b: u32| -> u32 {
                let y = (a & UPPER_MASK) | (b & LOWER_MASK);
                (y >> 1) ^ (MATRIX_A * (y & 1))
            };
            for kk in 0..N - M {
                self.mt[kk] = self.mt[kk + M] ^ twist(self.mt[kk], self.mt[kk + 1]);
            }
            for kk in N - M..N - 1 {
                self.mt[kk] = self.mt[kk + M - N] ^ twist(self.mt[kk], self.mt[kk + 1]);
            }
            self.mt[N - 1] = self.mt[M - 1] ^ twist(self.mt[N - 1], self.mt[0]);
            self.index = 0;
        }
        let mut y = self.mt[self.index];
        self.index += 1;
        y ^= y >> 11;
        y ^= (y << 7) & 0x9d2c_5680;
        y ^= (y << 15) & 0xefc6_0000;
        y ^= y >> 18;
        y
    }

    /// The next double in `[0, 1)`, exactly like `random.random()`.
    pub fn next_f64(&mut self) -> f64 {
        let a = (self.next_u32() >> 5) as f64;
        let b = (self.next_u32() >> 6) as f64;
        (a * 67108864.0 + b) * (1.0 / 9007199254740992.0)
    }

    /// `a + (b - a) * random()`, like `random.uniform`.
    pub fn uniform(&mut self, a: f64, b: f64) -> f64 {
        a + (b - a) * self.next_f64()
    }
}

#[cfg(test)]
mod tests {
    use super::Mt19937;

    #[test]
    fn the_stream_is_pythons() {
        // random.Random(0).random() three times, and random.Random(1).random()
        let mut rng = Mt19937::new(0);
        assert!((rng.next_f64() - 0.8444218515250481).abs() < 1e-15);
        assert!((rng.next_f64() - 0.7579544029403025).abs() < 1e-15);
        assert!((rng.next_f64() - 0.420571580830845).abs() < 1e-15);
        let mut rng = Mt19937::new(1);
        assert!((rng.next_f64() - 0.13436424411240122).abs() < 1e-15);
    }
}
